### PR TITLE
feat: add support for nested modules

### PR DESCRIPTION
**Pull request by tohrnii**
## Describe your changes
Addressing #983

## Checklist before requesting a review
- Repo forked and branch created from `next` according to naming convention.
- Commit messages and codestyle follow [conventions](./CONTRIBUTING.md).
- Relevant issues are linked in the PR description.
- Tests added for new functionality.
- Documentation/comments updated according to changes.

### CODE REVIEW

File: assembly/src/library/path.rs [hunk 1]
```diff
@@ -79,20 +83,27 @@ impl LibraryPath {
...
-        self.path
-            .split_once(Self::PATH_DELIM)
-            .expect("a valid library path must always have at least one component")
-            .0
+        if let Some((first, _)) = self.path.split_once(Self::PATH_DELIM) {
+            first
+        } else {
+            self.path.as_str()
+        }
     }
```
Comment (bobbinth) [inline assembly/src/library/path.rs:91]: Is this change needed to support `mod.masm` files in the root directory of a library? If so, I think we can disallow this.

File: assembly/src/library/path.rs [hunk 2]
```diff
@@ -79,20 +83,27 @@ impl LibraryPath {
...
-        self.path
-            .rsplit_once(Self::PATH_DELIM)
-            .expect("a valid library path must always have at least one component")
-            .1
+        if let Some((_, last)) = self.path.rsplit_once(Self::PATH_DELIM) {
+            last
+        } else {
+            self.path.as_str()
+        }
+    }
```
Comment (bobbinth) [inline assembly/src/library/path.rs:102]: Similar comment to the previous one.

File: assembly/src/procedures/mod.rs [hunk 3]
```diff
@@ -244,6 +246,14 @@ impl From<[u8; ProcedureId::SIZE]> for ProcedureId {
...
     fn from(path: &LibraryPath) -> Self {
+        let penultimate = path.num_components().saturating_sub(2);
+        let path =
+            if path.num_components() >= 2 && path.at_pos(penultimate) == Some(LibraryPath::MOD) {
+                path.remove_at(penultimate)
+                    .expect("could not remove penultimate path component")
+            } else {
+                path.clone()
+            };
         ProcedureId::new(path)
```
Comment (bobbinth) [inline assembly/src/procedures/mod.rs:257]: I'd prefer to avoid path manipulations here. Ideally, the changes should not affect anything outside of `/src/library` modules. Basically, by the time we are here, module paths should have been adjusted so that `x/y/mod.masm` becomes `x::y`

File: assembly/src/library/path.rs
Comment (bobbinth) [inline assembly/src/library/path.rs:1]: We should probably add validation checks to make sure paths like `x::y::mod::z` or `x::y::mod` are not possible.

**Review by bobbinth — CHANGES_REQUESTED** [review]
Thank you! Looks good! I left some comments inline. Basically, I'm thinking we can change the approach so that all path manipulations are done at the time when we build libraries and everything else works as is. Basically, as we traverse directory tree, if we see something like `x/y/mod.masm` we transform this into library path `x::y`.

We should also probably add a check to the library construction logic to make sure that we get an error if there a file with the same name as the directory. For example:
```
/x/y/mod.masm
/x/y.masm
```
The above should fail because having `y.masm` conflicts with `y/mod.masm`.

File: assembly/src/library/path.rs [hunk 4]
```diff
@@ -332,6 +334,8 @@ fn validate_path_len(path: &str) -> Result<(), PathError> {
 fn validate_component(component: &str) -> Result<(), PathError> {
     if component.is_empty() {
         Err(PathError::EmptyComponent)
+    } else if component == LibraryPath::MOD {
+        Err(PathError::component_invalid_name(component))
```
Comment (bobbinth) [inline assembly/src/library/path.rs:338]: Now that I think about it, I wonder if we need this check at all. For `.masl` libraries, having `mod` in the path would not be possible by construction - but maybe for some other libraries it could be valid (e.g., if they adopt a different strategy for defining parent modules).

File: assembly/src/library/path.rs [hunk 5]
```diff
@@ -32,6 +32,8 @@ impl LibraryPath {
     /// Path for an executable module.
     pub const EXEC_PATH: &str = "#exec";
 
+    pub const MOD: &str = "mod";
```
Comment (bobbinth) [inline assembly/src/library/path.rs:35]: If we remove the check mentioned in the previous comment, this would probably need to go into `MaslLibrary`.

**Review by bobbinth — CHANGES_REQUESTED** [review]
Looks good! Thank you! I left a couple of small comments inline.

One thing I'm wondering is whether there is a good way to test this somehow. Maybe we could add some useful function under `std::math` or something like that?

File: assembly/src/library/path.rs (hunk 4 above)
Comment (tohrnii) [reply assembly/src/library/path.rs:338]: I think maybe not making assumptions about how the library was serialized and keeping this check is probably better since technically we can have a `.masl` library with that path if we don't use the provided bundling logic because other than this the only place we check this is during bundling (for e.g. someone doing their own serialization for some reason).

**Comment by tohrnii** [comment]
> One thing I'm wondering is whether there is a good way to test this somehow. Maybe we could add some useful function under std::math or something like that?

One option is to re-export methods of the `fri` module for example from `mod.masm`?

**Comment by bobbinth** [comment]
> One option is to re-export methods of the `fri` or `ecdsa` module from `mod.masm`?

Good idea! Let's do this: re-export `std::crypto::stark::verifier::verify` from `std::crypto::stark`.

File: assembly/src/library/path.rs (hunk 4 above)
Comment (bobbinth) [reply assembly/src/library/path.rs:338]: But would having `mod` in the path cause an issue of some sort?

Comment (tohrnii) [reply assembly/src/library/path.rs]: Yeah, you're right it shouldn't so I've removed the check. But maybe it's cleaner to disallow using the keyword in path in general?

**Comment by tohrnii** [comment]
One other question is if we should add a build step to copy the docs from `stdlib` crate to `docs` crate so that they are in sync.

File: assembly/src/library/path.rs (hunk 4 above)
Comment (bobbinth) [reply assembly/src/library/path.rs]: I'm not sure we should consider `mod` as a keyword in the the assembly language. It is more like a specific way `.masl` libraries work - but some other library bundle strategy may work differently and allow `mod` in the path.

**Comment by bobbinth** [comment]
> One other question is if we should add a build step to copy the docs from `stdlib` crate to `docs` crate so that they are in sync.

No, the formatting on auto-generated docs needs to be improved quite a bit.

**Review by bobbinth — APPROVED** [review]
All looks good! Thank you!
